linera-core/src/remote_node.rs

```diff
@@ -246,54 +242,16 @@ impl<N: ValidatorNode> RemoteNode<N> {
         }
     }
 
-    /// Returns the list of certificate hashes on the given chain in the given range of heights.
-    /// Returns an error if the number of hashes does not match the size of the range.
+    /// Downloads a range of certificates from the given chain.
     #[instrument(level = "trace")]
-    pub(crate) async fn fetch_sent_certificate_hashes(
+    pub async fn download_certificates_by_heights(
         &self,
         chain_id: ChainId,
-        range: BlockHeightRange,
-    ) -> Result<Vec<CryptoHash>, NodeError> {
-        let query =
-            ChainInfoQuery::new(chain_id).with_sent_certificate_hashes_in_range(range.clone());
-        let response = self.handle_chain_info_query(query).await?;
-        let hashes = response.requested_sent_certificate_hashes;
-
-        if range
-            .limit
-            .is_some_and(|limit| hashes.len() as u64 != limit)
-        {
-            warn!(
-                ?range,
-                received_num = hashes.len(),
-                "Validator sent invalid number of certificate hashes."
-            );
-            return Err(NodeError::InvalidChainInfoResponse);
-        }
-        Ok(hashes)
-    }
-
-    #[instrument(level = "trace")]
-    pub async fn download_certificates(
-        &self,
-        hashes: Vec<CryptoHash>,
+        heights: Vec<BlockHeight>,
     ) -> Result<Vec<ConfirmedBlockCertificate>, NodeError> {
-        if hashes.is_empty() {
-            return Ok(Vec::new());
-        }
-        let certificates = self.node.download_certificates(hashes.clone()).await?;
-        let returned = certificates
-            .iter()
-            .map(ConfirmedBlockCertificate::hash)
-            .collect();
-        ensure!(
-            returned == hashes,
-            NodeError::UnexpectedCertificates {
-                returned,
-                requested: hashes
-            }
-        );
-        Ok(certificates)
+        self.node
+            .download_certificates_by_heights(chain_id, heights)
```

Comment:
Is there a reason we can't put the functionality right here instead of adding the layer of indirection?

Reply:
This code is in the client. (It's the client's handle of a remote node connection.) So that would create a lot of client-proxy round-trips in addition to the proxy-server round-trips which should be much faster.

Reply:
Yes. More importantly – if that wasn't clear – is that we make ONE query to a validator vs multiple queries. Now the "multiple queries" happens between validator's proxy and worker.